torch/storage.py

```diff
@@ -163,3 +139,288 @@ def _load_from_bytes(b):
+                # methods?
+                self.storage = storage.storage
+
+            else:
+                assert isinstance(storage, eval(storage.__module__ + ".ByteStorage"))
+                self.storage = storage
+        else:
+            assert type(self) != TypedStorage, (
+                "Calling __init__ this way is only supported in TypedStorage's "
+                "child classes. TypedStorage can only be directly instantiated "
+                "when kwargs 'wrap_storage' and 'dtype' are given.")
+
+            # TODO: Still need to override some of the constructor behavior.
+            # For instance, when we do something like
+            # `torch.FloatStorage([0, 1, 2])`, we'll end up calling
+            # `torch.ByteStorage([0, 1, 2])` here, which is not correct,
+            # since we want the storage to be filled with 3 float values
+            # [0, 1, 2]
+            self.storage = eval(self.__module__ + ".ByteStorage")(*args, **kwargs)
+
+    def _untyped(self):
+        return self.storage
+
+    def _new_wrapped_storage(self, byte_storage):
+        module = eval(byte_storage.__module__)
+        assert type(byte_storage) == module.ByteStorage
+
+        if type(self) == TypedStorage:
+            return TypedStorage(wrap_storage=byte_storage, dtype=self.dtype)
+        else:
+            # NOTE: We need to use the module of byte_storage in case self's
+            # module is different. e.g. if self is on CPU and byte_storage
+            # is on CUDA, and vice versa
+            return getattr(module, type(self).__name__)(wrap_storage=byte_storage)
+
+    def __len__(self):
+        return self.storage.nbytes() // self.element_size()
+
+    def _maybe_wrap_index(self, idx):
+        if type(idx) != int:
+            raise TypeError(
+                f"can't index a {type(self)} with {type(idx)}")
+
+        if idx is None:
+            return 0
+
+        else:
+            if (idx >= self.size()) or (idx < -self.size()):
+                raise IndexError(
+                    f'index {idx} out of range for storage of size {self.size()}')
+            return idx % self.size()
+
+
+    def __setitem__(self, idx, value):
+        # TODO: Seems bad to use a tensor inside a storage method
+        tmp_tensor = torch.tensor(self.storage, device=self.device, dtype=self.dtype)
+        tmp_tensor[idx] = value
+
+    def __getitem__(self, idx):
+        if isinstance(idx, slice):
+            # Need to wrap start and stop, then multiply each slice param
+            # by element_size
+            start = self.element_size() * self._maybe_wrap_index(idx.start)
+            stop = self.element_size() * self._maybe_wrap_index(idx.stop)
+            
+            if idx.step is not None and idx.step != 1:
+                raise RuntimeError((
+                    f'Trying to slice with a step of {idx.step}, but '
+                    'only a step of 1 is supported'))
+
+            byte_slice = slice(start, stop, 1)
```

Comment:
I'm a little wary about converting a slice intended for a typed storage into a matching slice for a ByteStorage. But I don't know if there's any other way to do this

Reply:
you can also simulate slicing using tensor here, right?

Reply:
At the moment, I'm not sure if it's possible to use tensor for this. In current pytorch, when we index a storage with a slice, we get a new storage object that points to the same data, whose size is only as large as the slice:

```python
>>> s = torch.arange(10).storage()
>>> s.data_ptr()
94174994880832
>>> s[:4].data_ptr()
94174994880832
>>> s[:4]
 0
 1
 2
 3
[torch.LongStorage of size 4]
```

But when we index a tensor with a slice, the sliced tensor's storage remains the full original size:

```python
>>> a = torch.arange(10)
>>> a[:4]
tensor([0, 1, 2, 3])
>>> a[:4].storage()
 0
 1
 2
 3
 4
 5
 6
 7
 8
 9
[torch.LongStorage of size 10]
```

I could potentially add a new tensor method that creates a new shortened storage object over only the range that the tensor touches. What do you think?

Reply:
Nevermind, I forgot storage slicing produces storage views. My updated recommendation is that we just delete this functionality :>